lib/client/configurator/docker_linux.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package configurator
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/gravitational/trace"
+)
+
+type dockerConfigurator struct {
+	// debug enables verbose logging mode for debugging.
+	debug bool
+}
+
+// Configure configures local Docker with client key/certificate specified
+// in the provided configuration.
+func (c *dockerConfigurator) Configure(config Config) error {
+	if !hasDocker() {
+		log.Debug("Will not configure Docker registy: docker not available.")
+		return nil
+	}
+	// Docker configuration works by symlinking client key/certificate to
+	// a location inside /etc/docker/certs.d which requires root privileges
+	// so if we're not running as root, invoke a subcommand as root that
+	// will relaunch this configurator and perform the symlinking below.
+	if os.Geteuid() != 0 {
+		return c.runAsRoot(config)
+	}
+	// Ensure /etc/docker/certs.d/<proxy> directory exists.
+	certsDir := filepath.Join(DockerCerts, config.ProxyAddress)
+	if err := os.MkdirAll(certsDir, 0755); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	// Symlink user's key/certificate to /etc/docker/certs.d/<proxy>.
+	symlinks := c.getSymlinks(config)
+	if err := ensureSymlinks(symlinks); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// IsConfigured returns true if the local Docker is already configured with
+// the specified client key/certificate.
+func (c *dockerConfigurator) IsConfigured(config Config) (bool, error) {
+	symlinks := c.getSymlinks(config)
+	return verifySymlinks(symlinks)
+}
+
+// getSymlinks returns a map of symlinks that need to be configured in order
+// to let local Docker access registry provided by the proxy.
+func (c *dockerConfigurator) getSymlinks(config Config) map[string]string {
+	certsDir := filepath.Join(DockerCerts, config.ProxyAddress)
```

Comment:
Since `config` can be attacker controlled, `config.ProxyAddress` can be something like `../../../some/other/dir`, I would check `certsDir` after the join that the base directory is what you expect.